### PR TITLE
GKE Driver Functional

### PR DESCRIPTION
**Pull request by andymaheshw**
### Description

Creating the google container driver
### Status

- work in progress

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [X] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

**Comment by andymaheshw** [comment]
Working on getting the test to run, and cleaning up the docs. @tonybaloney this is just a quick update to let you know that it is moving along.

File: docs/container/drivers/joyent.rst [hunk 1]
```diff
@@ -17,7 +17,7 @@ Instantiating the driver
 Download the script::
 
     curl -O https://raw.githubusercontent.com/joyent/sdc-docker/master/tools/sdc-docker-setup.sh
-    
+
```
Comment (tonybaloney) [inline docs/container/drivers/joyent.rst]: if you wouldn't mind just undoing the changes to the joyent file, `git checkout trunk docs/container/drivers/joyent.rst`

File: docs/container/drivers/gke.rst [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    ****
+
+Now execute the script, substituting the correct values::
+
+    ****
+
+This should output something similar to the following::
+
+    Setting up Docker client for SDC using:
```
Comment (tonybaloney) [inline docs/container/drivers/gke.rst]: Lines 23-24 need updating with some google specific info

File: libcloud/container/drivers/gke.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
+from libcloud.common.google import GoogleOAuth2Credential
+
+from libcloud.container.drivers.docker import (DockerContainerDriver,
+                                               DockerConnection)
+
+
+class GKEContainerDriver(DockerContainerDriver):
```
Comment (tonybaloney) [inline libcloud/container/drivers/gke.py]: Will you inherit from the KubernetesContainerDriver class? Also check out the container driver type enum. You want the clustered one.

**Review by tonybaloney — COMMENTED** [review]
Good start, looking forward to seeing this working 👍 thanks for contributing!

**Comment by andymaheshw** [comment]
Thanks, going to work on this tomorrow!

File: libcloud/container/drivers/gke.py (hunk 3 above)
Comment (andymaheshw) [reply libcloud/container/drivers/gke.py]: Hi @tonybaloney What did you mean by this? I am a bit unclear.

Comment (tonybaloney) [reply libcloud/container/drivers/gke.py]: http://libcloud.readthedocs.io/en/latest/container/index.html#cluster-suppport

**Comment by andymaheshw** [comment]
@tonybaloney This was a fun challenge, I can't wait to build more into it. Let me know what you think!

File: docs/container/drivers/gke.rst [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+1. Getting Driver with Service Account authentication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. literalinclude:: /examples/compute/gce/gce_service_account.py
+
+
+API Docs
+--------
+
+.. autoclass:: libcloud.container.drivers.joyent.GKEContainerDriver
```
Comment (tonybaloney) [inline docs/container/drivers/gke.rst]: this module path contains joyent, you probably mean gke

File: docs/examples/container/gke/instantiate_driver.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from libcloud.container.types import Provider
+from libcloud.container.providers import get_driver
+
+cls = get_driver(Provider.GKE)
+
+conn = cls('testaccount-XXX@testproject.iam.gserviceaccount.com',
+           'libcloud.json', project='testproject')
+
+conn.list_images()
```
Comment (tonybaloney) [inline docs/examples/container/gke/instantiate_driver.py]: choose a more container-specific example method

File: libcloud/container/drivers/gke.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+class GKEResponse(GoogleResponse):
+    pass
+
+
+class GKEConnection(GoogleBaseConnection):
+    """
+    Connection class for the GKE driver.
+
+    GKEConnection extends :class:`google.GoogleBaseConnection` for 2 reasons:
```
Comment (tonybaloney) [inline libcloud/container/drivers/gke.py]: that's 3 reasons ;-)

File: libcloud/container/drivers/gke.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    connectionCls = GKEConnection
+    api_name = 'google'
+    name = "Google Container Engine"
+    type = Provider.GKE
+    website = 'https://container.googleapis.com'
+    supports_clusters = True
+
+    AUTH_URL = "https://container.googleapis.com/auth/"
+
+    BACKEND_SERVICE_PROTOCOLS = ['HTTP', 'HTTPS', 'HTTP2', 'TCP', 'SSL']
```
Comment (tonybaloney) [inline libcloud/container/drivers/gke.py]: is this used anywhere or a copy+paste from another module?

File: libcloud/container/drivers/kubernetes.py [hunk 8]
```diff
@@ -123,26 +123,29 @@ def __init__(self, key=None, secret=None, secure=False, host='localhost',
 
         :return: ``None``
         """
+        print(key, secret)
```
Comment (tonybaloney) [inline libcloud/container/drivers/kubernetes.py]: remove this leftover

File: libcloud/test/container/test_gke.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        kwargs['auth_type'] = 'IA'
+        kwargs['datacenter'] = self.datacenter
+        self.driver = GKEContainerDriver(*GKE_PARAMS, **kwargs)
+
+    def test_list_images_response(self):
+        pass
+
+    def test_server_config(self):
+        zones = self.driver.get_server_config()
+        print(zones)
```
Comment (tonybaloney) [inline libcloud/test/container/test_gke.py]: remove this one too

File: libcloud/test/container/test_gke.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    def _get_method_name(self, type, use_param, qs, path):
+        api_path = '/%s' % API_VERSION
+        project_path = '/projects/%s' % GKE_KEYWORD_PARAMS['project']
+        path = path.replace(api_path, '')
+        # This replace is separate, since there is a call with a different
+        # project name
+        path = path.replace(project_path, '')
+        # The path to get project information is the base path, so use a fake
+        # '/project' path instead
+        print("path", path)
```
Comment (tonybaloney) [inline libcloud/test/container/test_gke.py]: and this one.

File: libcloud/container/drivers/gke.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        self.base_path = '/%s/projects/%s' % (API_VERSION, self.project)
+        self.website = GKEContainerDriver.website
+
+    def _ex_connection_class_kwargs(self):
+        return {'auth_type': self.auth_type,
+                'project': self.project,
+                'scopes': self.scopes,
+                'credential_file': self.credential_file}
+
+    def list_images(self, zone=None):
```
Comment (tonybaloney) [inline libcloud/container/drivers/gke.py]: you need to document what this is and what the zone param is for. Also, it should be named `ex_zone`, to indicate it is a non-standard argument.

File: libcloud/container/drivers/gke.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        """
+        """
+        request = "/zones/%s/clusters" % (zone)
+        if zone is None:
+            request = "/zones/clusters"
+
+        response = self.connection.request(request, method='GET').object
+        return response
+
+    def get_server_config(self, zone=None):
```
Comment (tonybaloney) [inline libcloud/container/drivers/gke.py]: same as above

**Review by tonybaloney — CHANGES_REQUESTED** [review]

**Comment by andymaheshw** [comment]
@tonybaloney not sure about `list_clusters()` for the method name, let me know :) Thanks for the fast comments.

File: libcloud/compute/drivers/gce.py [hunk 13]
```diff
@@ -2750,6 +2750,7 @@ def ex_list_zones(self):
         :return: A list of zone objects.
         :rtype: ``list`` of :class:`GCEZone`
         """
+        print("WHEN DO WE DO THIS")
```
Comment (tonybaloney) [inline libcloud/compute/drivers/gce.py]: you did it again

**Comment by tonybaloney** [comment]
now if you can review the lint and pylint output on travis https://travis-ci.org/apache/libcloud/builds/244486577?utm_source=github_status&utm_medium=notification

**Comment by andymaheshw** [comment]
ok! all set, I have triple checked for print statements.

**Comment by tonybaloney** [comment]
👍  LGTM

**Comment by tonybaloney** [comment]
merged

**Pull request merged**